False positive in NoWhiteSpaceAfter

hi,

starting with ~~8.45~~ 8.44, checkstyle generates a false positive for this code

```
public class Demo {

    Object o;

    void some() {
        Object oo = new Object[4];
        this.o = ((Object[]) oo)[1];
    }
}
```
This leads to message like "There is a whitespace after 'o'"

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="NoWhitespaceAfter">
    </module>
  </module>
</module>
```
produces this output using commandline / latest release 10.3.4
```
Starting audit...
[ERROR] ...\Demo.java:7:33: 'o' is followed by whitespace. [NoWhitespaceAfter]
Audit done.
Checkstyle ends with 1 errors.
```

Above jave code (Demo.java) compiles without error or warning

Using this config:
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="NoWhitespaceAfter">
    </module>
  </module>
</module>
```
produces this output using commandline / latest release 10.3.4
```
Starting audit...
[ERROR] ...\Demo.java:7:33: 'o' is followed by whitespace. [NoWhitespaceAfter]
Audit done.
Checkstyle ends with 1 errors.
```

Strange, it seems to be related to the combination of redundant `this` and casting to array type with access:

```

➜  src cat Test.java                                              
public class Test {

    Object o;

    void m1() {
        Object oo = new Object[4];
        this.o = ((Object[]) oo)[1]; // violation is here
    }

    void m2() {
        Object oo = new Object[4];
        o = ((Object[]) oo)[1];
    }

    void m3() {
        Object oo = new Object[4];
        this.o = new Object();
    }

    void m4() {
        Object oo = new Object[4];
        this.o = (int) oo;
    }

    void m5() {
        Object oo = new Object[4];
        this.o = ((int[][]) new Object())[1];
    }
}%                                                                                                             

 ➜  src cat config.xml                                             
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="NoWhitespaceAfter">
    </module>
  </module>
</module>%  
                                                                                                    
➜  src javac Test.java                                            
➜  src java -jar checkstyle-10.3.4-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/checkstyle_tester/src/Test.java:7:33: 'o' is followed by whitespace. [NoWhitespaceAfter]
[ERROR] /home/nick/IdeaProjects/checkstyle_tester/src/Test.java:27:42: 'o' is followed by whitespace. [NoWhitespaceAfter]
Audit done.
Checkstyle ends with 2 errors.

```

--- a/.ci/checker-framework-suppressions/checker-nullness-optional-interning-suppressions.xml
+++ b/.ci/checker-framework-suppressions/checker-nullness-optional-interning-suppressions.xml
@@ -4521,13 +4521,6 @@
   </checkerFrameworkError>
 
   <checkerFrameworkError unstable="false">
-    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java</fileName>
-    <specifier>introduce.eliminate</specifier>
-    <message>It is bad style to create an Optional just to chain methods to get a value.</message>
-    <lineContent>return Optional.ofNullable(referencedMemberDot).orElse(referencedClassDot);</lineContent>
-  </checkerFrameworkError>
-
-  <checkerFrameworkError unstable="false">
     <fileName>src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/ParenPadCheck.java</fileName>
     <specifier>not.interned</specifier>
     <message>attempting to use a non-@Interned comparison operand</message>

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java
@@ -506,13 +506,12 @@ public class NoWhitespaceAfterCheck extends AbstractCheck {
         final DetailAST previousElement;
         final DetailAST ident = getIdentLastToken(parent.getParent());
         final DetailAST lastTypeNode = getTypeLastNode(ast);
-        final boolean isTypeCast = parent.getParent().getType() == TokenTypes.TYPECAST;
         // sometimes there are ident-less sentences
         // i.e. "(Object[]) null", but in casual case should be
         // checked whether ident or lastTypeNode has preceding position
         // determining if it is java style or C style
 
-        if (ident == null || isTypeCast || ident.getLineNo() > ast.getLineNo()) {
+        if (ident == null || ident.getLineNo() > ast.getLineNo()) {
             previousElement = lastTypeNode;
         }
         else if (ident.getLineNo() < ast.getLineNo()) {
@@ -544,9 +543,9 @@ public class NoWhitespaceAfterCheck extends AbstractCheck {
      */
     private static DetailAST getIdentLastToken(DetailAST ast) {
         final DetailAST result;
-        final DetailAST dot = getPrecedingDot(ast);
+        final Optional<DetailAST> dot = getPrecedingDot(ast);
         // method call case
-        if (dot == null || ast.getFirstChild().getType() == TokenTypes.METHOD_CALL) {
+        if (dot.isEmpty() || ast.getFirstChild().getType() == TokenTypes.METHOD_CALL) {
             final DetailAST methodCall = ast.findFirstToken(TokenTypes.METHOD_CALL);
             if (methodCall == null) {
                 result = ast.findFirstToken(TokenTypes.IDENT);
@@ -557,7 +556,7 @@ public class NoWhitespaceAfterCheck extends AbstractCheck {
         }
         // qualified name case
         else {
-            result = dot.getFirstChild().getNextSibling();
+            result = dot.get().getFirstChild().getNextSibling();
         }
         return result;
     }
@@ -569,11 +568,24 @@ public class NoWhitespaceAfterCheck extends AbstractCheck {
      * @param leftBracket the ast we are checking
      * @return dot preceding the left bracket
      */
-    private static DetailAST getPrecedingDot(DetailAST leftBracket) {
-        final DetailAST referencedClassDot =
-                leftBracket.getParent().findFirstToken(TokenTypes.DOT);
+    private static Optional<DetailAST> getPrecedingDot(DetailAST leftBracket) {
         final DetailAST referencedMemberDot = leftBracket.findFirstToken(TokenTypes.DOT);
-        return Optional.ofNullable(referencedMemberDot).orElse(referencedClassDot);
+        final Optional<DetailAST> result = Optional.ofNullable(referencedMemberDot);
+        return result.or(() -> getReferencedClassDot(leftBracket));
+    }
 
+    /**
+     * Gets the dot preceding a class reference.
+     *
+     * @param leftBracket the ast we are checking
+     * @return dot preceding the left bracket
+     */
+    private static Optional<DetailAST> getReferencedClassDot(DetailAST leftBracket) {
+        final DetailAST parent = leftBracket.getParent();
+        Optional<DetailAST> classDot = Optional.empty();
+        if (parent.getType() != TokenTypes.ASSIGN) {
+            classDot = Optional.ofNullable(parent.findFirstToken(TokenTypes.DOT));
+        }
+        return classDot;
     }
 }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheckTest.java
@@ -64,6 +64,13 @@ public class NoWhitespaceAfterCheckTest
     }
 
     @Test
+    public void testAssignment() throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+                getPath("InputNoWhitespaceAfterTestAssignment.java"), expected);
+    }
+
+    @Test
     public void testDotAllowLineBreaks() throws Exception {
         final String[] expected = {
             "9:13: " + getCheckMessage(MSG_KEY, "."),

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterTestAssignment.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterTestAssignment.java
@@ -1,0 +1,41 @@
+/*
+NoWhitespaceAfter
+allowLineBreaks = false
+tokens = (default)ARRAY_INIT, AT, INC, DEC, UNARY_MINUS, UNARY_PLUS, BNOT, LNOT, \
+         DOT, ARRAY_DECLARATOR, INDEX_OP
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.whitespace.nowhitespaceafter; // ^ 2 violations above
+
+public class InputNoWhitespaceAfterTestAssignment {
+
+    Object o;
+    static boolean b = true;
+
+    void some() {
+        Object oo = new Object[4];
+        Object[] oo2 = new Object[4];
+        this.o = ((Object[]) oo)[1]; // ok
+        this.o = ((java.lang.Object[]) oo)[1]; // ok
+        this.o = oo2[1];
+        QualifiedAssignment.o1 = ((Object[]) oo)[1]; // ok
+        QualifiedAssignment.o1 = ((java.lang.Object[]) oo)[1]; // ok
+        QualifiedAssignment.o1 = oo2[1];
+        QualifiedAssignment qa1 = null;
+        QualifiedAssignment[] qa2 = null;
+        int idx = 0;
+        (qa1 = (QualifiedAssignment)qa2[idx]).o1 = (new QualifiedAssignment[idx][idx][idx])[idx];
+        (b ? (new QualifiedAssignment().q1 = new QualifiedAssignment()) :
+            (QualifiedAssignment)(new QualifiedAssignment().q1 = new QualifiedAssignment())).q1 =
+            (new QualifiedAssignment[new QualifiedAssignment().idx = (QualifiedAssignment.idx =
+                QualifiedAssignment.idx)])[QualifiedAssignment.idx];
+    }
+}
+
+class QualifiedAssignment {
+    static Object o1;
+    static QualifiedAssignment q1;
+    static int idx = 1;
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterTestDefault.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterTestDefault.java
@@ -192,7 +192,7 @@ class InputNoWhitespaceAfterTestDefault
     /** bug 806243 (NoWhitespaceBeforeCheck violation for anonymous inner class) */
     void bug806243()
     {
-        Object o = new InputNoWhitespaceAfterTestDefault() {
+        Object o = new InputNoWhitespaceAfterTestAssignment() {
             private int j ;
             //           ^ whitespace
         };

